guides/technical-guides/test-driven-development.md

```diff
@@ -0,0 +1,105 @@
+# Test Driven Development (TDD)
+
+It is expected that where feasible and appropriate, all projects will have
+relevant tests and test automation to accompany the functional elements of the
+project.
+
+One great way to ensure this practice is adhered to is through the medium of
+**Test Driven Development**, commonly referred to as **TDD**.
+
+## What is TDD?
+
+TDD is a method of development whereby (in an _ideal_ world) unit tests are
+written before the application code of the project.
+
+The ideal flow of TDD is as follows:
+
+1. Write a failing test
+2. Write functional code to make the test pass
+3. Yay, we have passing tests!
+
+## Why TDD?
+
+The advantage of writing code in a TDD way is that one can set out exactly what
+should make a test pass and what shouldn't.
+
+Writing the application code after the test means that the bare minimum code is
+written to fulfil the requirement for a particular test. Additionally, the
+theory goes that writing the test beforehand means the test shouldn't be forced
+to pass by writing complex test logic.
+
+This helps to reduce the complexity of individual methods and can reduce scope
+creep for particular requirements.
+
+TDD also helps to reduce bugs in the code as individual methods are tested only
+for what they are written to achieve. This is great because the testing is
+low-level, meaning that differing conditions at runtime _probably_ won't hide
+other issues.
+
+## Example (in Python 🐍)
+
+_Assume the relevant `unittest` packages are imported and setup._
+
+1. Write a test that should pass, but fails because the method isn't implemented
+   yet.
+
+   `test_calculator.py`
+
+   ```python
+   def test_calculator_add(self):
```

Comment:
As this example uses `unittest` can you add a class, please?

```python
import unittest

class TestCalculator(unittest.TestCase):
    def test_calculator_add(self):
        self.assertEquals(Calculator.add(2, 2), 4)
```

Reply:
Yeah done